File: Managed/UnrealSharp/UnrealSharp.Core/Interop/Bind_UCSManager.cs
using UnrealSharp.Binds;

namespace UnrealSharp.Core.Interop;

[NativeCallbacks]
public static unsafe partial class Bind_UCSManager
{
    public static delegate* unmanaged<IntPtr, IntPtr> FindManagedObject;
    public static delegate* unmanaged<IntPtr, IntPtr, IntPtr> FindOrCreateManagedInterfaceWrapper;
    public static delegate* unmanaged<IntPtr> GetCurrentWorldContext;
    public static delegate* unmanaged<IntPtr> GetCurrentWorldPtr;
    public static delegate* unmanaged<IntPtr, void> PushWorldContext;
    public static delegate* unmanaged<void> PopWorldContext;
    
    public static UnrealSharpObject? WorldContextObject
    {
        get
        {
            IntPtr worldContextObject = CallGetCurrentWorldContext();
            if (worldContextObject == IntPtr.Zero)
            {
                return null;
            }

            IntPtr handle = CallFindManagedObject(worldContextObject);
            return GCHandleUtilities.GetObjectFromHandlePtr<UnrealSharpObject>(handle);
        }
    }
}

File: Source/UnrealSharpCore/Private/Binds/Bind_UCSManager.cpp
#include "CSManager.h"

DECLARE_UNREALSHARP_BINDER(Bind_UCSManager)
{
	void* FindManagedObject(UObject* Object)
	{
		return UCSManager::Get().FindManagedObject(Object);
	}

	void* FindOrCreateManagedInterfaceWrapper(UObject* Object, UClass* NativeClass)
	{
		return UCSManager::Get().FindManagedInterfaceWrapper(Object, NativeClass);
	}

	void* GetCurrentWorldContext()
	{
		void* WorldContext = UCSManager::Get().GetCurrentWorldContext();
		return WorldContext;
	}

	void* GetCurrentWorldPtr()
	{
		return UCSManager::Get().GetCurrentWorldContext();
	}

	void PushWorldContext(UObject* WorldContextObject)
	{
		UCSManager::Get().PushWorldContext(WorldContextObject, false);
	}

	void PopWorldContext()
	{
		UCSManager::Get().PopWorldContext();
	}
	
	BIND_UNREALSHARP_FUNCTION(FindManagedObject)
	BIND_UNREALSHARP_FUNCTION(FindOrCreateManagedInterfaceWrapper)
	BIND_UNREALSHARP_FUNCTION(GetCurrentWorldContext)
	BIND_UNREALSHARP_FUNCTION(GetCurrentWorldPtr)
	BIND_UNREALSHARP_FUNCTION(PushWorldContext)
	BIND_UNREALSHARP_FUNCTION(PopWorldContext)
}
